keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/AgpVersionHandler.kt

```diff
@@ -0,0 +1,158 @@
+ */
+// TODO Can't use the newer `com.android.Version` until AGP 3.6.0+
+@file:Suppress("deprecation")
+package com.slack.keeper
+
+import com.android.build.api.transform.Transform
+import com.android.build.gradle.internal.pipeline.TransformTask
+import com.android.builder.model.Version.ANDROID_GRADLE_PLUGIN_VERSION
+import com.google.auto.service.AutoService
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.Directory
+import org.gradle.api.provider.Provider
+import org.gradle.kotlin.dsl.withType
+import org.gradle.util.VersionNumber
+import java.lang.reflect.Field
+import java.util.Locale
+import kotlin.reflect.KProperty1
+import kotlin.reflect.full.memberProperties
+
+internal val AGP_VERSION get() = VersionNumber.parse(ANDROID_GRADLE_PLUGIN_VERSION)
+
+interface AgpVersionHandler {
+
+  companion object {
+    fun getInstance(): AgpVersionHandler = lazyInstance
+
+    private val lazyInstance by lazy {
+      createHandler()
+    }
+
+    /** Creates a new patcher for the given environment. */
+    private fun createHandler(): AgpVersionHandler {
+      /** Known patchers, listed in order of preference. */
+      // TODO(zsweers) properly use a ServiceLoader for these?
+      val handlers = listOf(Agp35xPatcher(), Agp36xPatcher())
+
+      val baseVersion = AGP_VERSION.baseVersion
+      return handlers.filter { baseVersion >= it.minVersion }
+          .maxBy(AgpVersionHandler::minVersion)
+          ?: error(
+              "$TAG No applicable proguard task patchers found for this version of AGP ($AGP_VERSION). Please file a bug report with your AGP version")
+    }
+  }
+
+  /** Minimum AGP version for this patcher. */
+  val minVersion: VersionNumber
+
+  /** Returns the expected minifier name (i.e. 'R8' or 'Proguard') for the givven [project]. */
+  fun expectedMinifier(project: Project): String
+
+  /** Returns the interpolated task name for given [minifier] (i.e. 'R8' or 'Proguard') and variant. */
+  fun interpolatedTaskName(minifier: String, variant: String): String
+
+  /** Patches the provided [prop] into the target available proguard task. */
+  fun applyGeneratedRules(
+      project: Project,
+      extension: KeeperExtension,
+      prop: Provider<Directory>
+  )
+}
+
+@AutoService(AgpVersionHandler::class)
+class Agp35xPatcher : AgpVersionHandler {
+  private val proguardConfigurable: Class<out Transform> by lazy {
+    @Suppress("UNCHECKED_CAST")
+    Class.forName(
+        "com.android.build.gradle.internal.transforms.ProguardConfigurable") as Class<out Transform>
+  }
+
+  private val configurationFilesField: Field by lazy {
+    proguardConfigurable.getDeclaredField("configurationFiles").apply {
+      isAccessible = true
+    }
+  }
+
+  override val minVersion: VersionNumber = VersionNumber.parse("3.5.0")
+
+  override fun expectedMinifier(project: Project): String {
+    // On <3.6, proguard is the default
```

Comment:
I thought R8 was the default starting with 3.4.0: https://developer.android.com/studio/releases/gradle-plugin#behavior_changes

Reply:
Oh snap, good catch

Reply:
00dedf0